### PR TITLE
Add use_author()

### PR DESCRIPTION
**Pull request by avalcarcel9**
I talked with @jennybc at useR! about adding functionality to easily add or adapt authors in the Authors@R field in the description. Essentially the same functionality and use as `use_package()`. I think expert users add their information to the .Rprofile but more novice developers still add manually and often miss the field in `use_description()`. This could be useful for adding an author after package development as well.

I've never contributed to a tidyverse package so I'm not sure of your conventions. I tried to look up a few similar functions in the package and mimic those styles. I am happy to adapt the function or documentation to match your conventions, in the event of bugs, or to add functionality. Additionally, I did not provide tests for the new function. If you'd like to incorporate the `use_author()` function/feature I'm again happy to add tests or more documentation.

Possible TODOs:
TODO allow generic authors input like desc_set_authors(authors, file = ".", normalize = FALSE) for multiple author input
TODO create addin to prompt for author information
TODO create a snippet with author information for DESCRIPTION
TODO add tests
TODO check if the author input has partial overlap with one already listed and ask user if they would like to add the person as a new author or replace the existing author

### CODE REVIEW

**Comment by avalcarcel9** [comment]
Bug in line 89 *sorry*. `role` must be a string in the `desc::desc_add_author` function. I can fix and re-pull if you're interested in incorporating.

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: Can you use the styler package on all files or sections of files that you are touching? That will bring your formatting in line with our typical style. I'm not sure if it will deal with line length, but if it does not, note that I generally cap line length at 80 and use *Code > Reflow comment* on sections like this.

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: I think it's fine to keep this helper very narrow. Keeners can use the desc package directly.

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: These are interesting ideas, but let's set them aside at this time.

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: Definitely yes!

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: There should be some checks for obviously bad input here. The internal helper `is_string()` will be useful to you.

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: Are you sure you can't just load up the DESCRIPTION file, grab the current authors (possibly none, but that's going to be very very rare), and just call `desc::desc_add_author()`? What I'm really saying is: does there need to be so much conditional logic?

In general, I'm comfortable assuming that we start with our own output, i.e. that we (usethis) created the DESCRIPTION file. Or that it's not pathological in some way.

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: If this `if () {...} else if () {...}` stays, note that the `if()` block ends with a return, an "early return". So you don't need `else if() {...}`.  Bring all the code that follows back up to the top-level.

File: R/authors.R
Comment (jennybc) [inline R/authors.R]: Definitely don't worry about this case. Let's handle the 90% use case.

**Comment by avalcarcel9** [comment]
Thanks so much @jennybc ! I'll take a look at all of this and start working on incorporating the changes this week.

File: NAMESPACE
Comment (avalcarcel9) [reply NAMESPACE]: I've removed the @importFrom and simply used ::.

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: I used `styler` through the drop down in Addins. I've never used this package so I'm not super familiar with what it should convert things to or how to check it worked. I used the default `styler::tidyverse_styel()` on all the pages I worked with. I also used Code > Reflow comment to adjust the long roxygen header at the start. I tried to adjust lines to shorter lengths where possible.

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: I've removed this todo

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: I've adjusted these TODOs to 'long term' to keep them as ideas but allow us to move forward now.

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: I added tests! I have one test I'd like to add to check that a message is returned from the use_author function when the usethis default author (First, Last...) is in the description. I used ui_yeah() to prompt the message. It seems this is interactive and so expect_message() didn't work. Is there a way to add a test in that case?

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: I didn't manually check inputs since the inputs for my function are all passed to desc package functions. They do seems to have checks on inputs when someone is input incorrectly as a vector or numeric for example. I'm happy to add these checks if you prefer they be done within the usethis function directly.

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: I tried to incorporate this flow into this version of the function and get rid of the conditional logic. I still have some `if()` to assess what the state of the current description is in order to add an author or error the function.

I now error if the DESCRIPTION has 'Author:' rather than 'Authors@R' and send the user a message. I allow for addition of an author so long as 'Authors@R@' is a field (blank or pre-filled). I do not make changes if the author being input with `use_author()` is already in the DESCRIPTION. I ask the user if they would like to delete the usethis default author (First, Last, ...) if it is found in the DESCRIPTION.

I'd like to eventually change the error from finding 'Author' to prompting the user if they want to replace it with 'Authors@R' and continue adding and also prompt the user to add 'Authors@R' if it is missing and Authors was not found. Not sure if I'm over-thinking this though. I tried to stick with your advice and start by assuming the user used usethis or at least the generic DESCRIPTION generated when using File > New Directory > R Package...

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: Typo on my part. That is what I get for trying to code this while at useR. I've removed these conditionals and made sure to return(invisible only once at the end of the function.

File: R/authors.R
Comment (avalcarcel9) [reply R/authors.R]: Removed the TODO

**Comment by avalcarcel9** [comment]
@jennybc This version seems to be working fine. I made sure to fetch and merge updates between my old branch and the current devel `usethis` on gh. Some checks on travis are failing for the R devel version but they seem to be unrelated to this content. I tried to rigorously test out the function manually myself by assessing different test cases and prompting errors. I eventually made as many of these real tests so hopefully most major bugs are out but a grad student only has so much time in the day.

Looking forward to your feedback and I hope its ready to merge!

**Comment by GitHunter0** [comment]
Hey @avalcarcel9 , it is a helpful feature, any update here?

**Comment by avalcarcel9** [comment]
@GitHunter0 no updates! I was waiting to hear back about next steps to get this incorporated into the package.

**Comment by GitHunter0** [comment]
It is a pity, @avalcarcel9 , so let's wait for @jennybc

**Comment by jennybc** [comment]
I took a look recently and decided it needs work before it can be merged. For starters, copy editing of the docs and code restyling. And, much more vaguely, editing to be more consistent with existing code base.

**Comment by avalcarcel9** [comment]
@jennybc is there anything I can do to help with the edits and todos or does the remaining work need to be done with someone on your team? As a sort of independent contributor I'm not sure what you mean exactly by these comments and would need some more specific TODOs to help out. But I'm still more than happy to help in any way I can!

**Comment by jennybc** [comment]
The contributing guide for this repo contains good information, such as re: code style:

https://github.com/r-lib/usethis/blob/master/.github/CONTRIBUTING.md

You could also read the documentation carefully and copy edit it. I think there are some missing words / typos in places.

**Comment by jennybc** [comment]
Also the tests use a helper that has since been renamed to `create_local_package()`.

You generally need to merge in current main development from `master` of the main repo.

All of this is stuff I can do, but this is why finishing the PR isn't a quick and easy process.

**Comment by avalcarcel9** [comment]
@jennybc I completely understand the process takes a long time to rigorously validate and check. I started this process while in grad school and have since moved onto a new job during this crazy time. I'd still love the opportunity to help and as time permits would love to help out with these last tasks to get the code and functionality ready to integrate. That being said, I'm not sure when that time will pop up with my new responsibilities during this crazy time. So if you get to this and I haven't contributed feel free to finish it off. If not, hopefully at some point (soon-ish) I can go through the documentation for contributing, make the changes, and re-push changes. So sorry I can't help more immediatly.

**Comment by jennybc** [comment]
I refactored this a lot since various things have changed considerably since this story arc first started!

File: R/author.R
Comment (jennybc) [inline R/author.R:77]: usethis has "challenge" functions that make the user aware of a situation we consider to be suboptimal. They don't necessarily exit/error. But they often ask the user to say "I see your advice, I'm not taking it, YOLO" or give them the chance to back out and fix something.

File: R/author.R
Comment (jennybc) [inline R/author.R:99]: OTOH usethis's "check" functins do really check something and error if the condition does not hold.

File: R/author.R
Comment (jennybc) [inline R/author.R:108]: usethis's UI needs an overhaul (pivoting towards cli), which I'm not necessarily biting off right now. But when we touch errors, we're phasing in `usethis_abort()`.

Comment (ateucher) [reply R/author.R:107]: For the overhaul, will the `ui_oops()`, `ui_info()` etc. be part of that? I.e., `usethis_abort()` is just the start?

File: R/author.R
Comment (ateucher) [inline R/author.R:102]: Do you think checking the email would also be good? It's a pretty unique identifier.  I can see cases like "Andy Teucher" vs "Andrew Teucher" or "Jenny Bryan" vs "Jennifer Bryan" slipping through...

Could be (using `pmap()`):

```r
identical(x, given) && identical(y, family) || identical(z, email)
```

But maybe that's overkill

File: R/author.R
Comment (ateucher) [inline R/author.R:120]: Would it be an option to pull [this](https://github.com/r-lib/usethis/blob/c5f80632054589b01e6e754cc352c70e47753cab/R/description.R#L80-L88) out into a function that returns that list, and consult it for the default `Authors@R`?

File: R/author.R
Comment (ateucher) [inline R/author.R:133]: I think the output is fine, I think it's nice to see confirmation that it's doing something

**Review by ateucher — APPROVED** [review]
I think it looks great! I agree with the the 80% use case otherwise it would get overly complex. I had just a couple of suggestions/questions. Thanks for the explanations of conventions and development direction.

File: R/author.R
Comment (jennybc) [reply R/author.R:107]: Yeah it will be everything. Some of those function will certainly live on, but their guts will be modified to use cli.

File: R/author.R
Comment (jennybc) [reply R/author.R:41]: I thought about it and did take a look at author info for some existing CRAN packages. It feels pretty common to only include email for `cre` and I'm not sure there's a real purpose for including `email` for other contributors? Anyhow, that's how I ended up here.

So I guess I'm assuming that `use_author()` is going to be used mostly for adding folks in a non-`cre` role, i.e. that the maintainer is probably already set up. And that it's fine to omit `email` in this case.

This did inspire me to add `#' @inheritDotParams utils::person`, which leads to much better documentation of `...`, so that is a big improvement.

File: R/author.R
Comment (jennybc) [reply R/author.R:28]: Ah, I handled this above before I got to this point in the review.

File: R/author.R
Comment (jennybc) [reply R/author.R:15]: I'm not 100% convinced it's a net win to have lots of links from actual packages into the book, especially links to specific sections. It feels kind of brittle/perishable. So my current posture is that there has to be a really reason to do this. I did go back and read that passage and intentionally paraphrased it.

File: R/author.R
Comment (jennybc) [reply R/author.R:103]: Yeah, I convinced myself that this level of matching was overkill.

File: R/author.R
Comment (jennybc) [reply R/author.R]: Good idea. Done.

File: R/author.R
Comment (ateucher) [reply R/author.R:41]: That all makes sense to me!

File: R/author.R
Comment (ateucher) [reply R/author.R:15]: Yup, that makes sense

File: R/author.R
Comment (jennybc) [reply R/author.R:44]: I worked on the examples a bit, so use of `email` and `role` is demonstrated, but I think most calls will be more like the last one: `use_author("Charlie", "Brown")`.

Comment (ateucher) [reply R/author.R:44]: Yeah, I like that

**Pull request merged**
